### ISSUE TITLE
Auto-escape HTML input

### ISSUE THREAD

**Issue by foonathan** [jgm/cmark#176]
I'd like to have an input option that will treat HTML as literal text.

For example:

```
Some text with <tags>.
```

Will give only a text node with the literal contents `Some text with <tags>`, instead of the HTML node.

This option will basically auto escape all `<`,`>` and the like in regular text. I'm using cmark as part of a C++ documentation generator, where you'd often have template names and thus something that looks like HTML. Escaping all occurrences manually is not very convenient in that case.

**Comment by mity** [comment]
What about enclosing it in a codespan, e.g. `` `any_template<T>` ``?

1. It prevents any misinterpretation of any characters like some Markdown syntax constructs, including `<` and `>`, `*` (emphasis vs. pointer) etc.
2. And it's appropriate for the C++ templates as they actually are code.

**Comment by foonathan** [comment]
The exact problem appears with links. I use `[foo<T>]()` to link to the entity `foo<T>`. Extra characters there seem not desirable.

**Comment by jgm** [comment]
I would have thought that in documentation you'd want to use
a code span for things like `foo<T>`.  Thus,

    [`foo<T>`]()

which shouldn't create a problem.  Anyway, that's what
you're intended to do with code, which is general might
contain characters with some significance in Markdown.

You're asking for a special-purpose flag that allows you
to use `foo<T>` without putting it in code backticks.
But think of all the other bits of code you might want to
document that would cause similar problems if not put
in code backticks.  C pointers, for example, using `*`.
Best just always to use code backticks for code, as
that was how Markdown was designed.

**Comment by foonathan** [comment]
Using `[foo<T>]` is more typing though. And there are other cases where putting a code span would complicate the implementation.

If you don't want to support it, I have a workaround. But I think there is some use for this option - for example if you intend to output Latex where HTML is just ignored. Then the user can't reasonable input it, so it might jus be treated as text.

**Comment by nwellnhof** [comment]
This is a feature I'd use for the Apache Clownfish documentation generator as well. As @foonathan mentioned, if a single output format doesn't support HTML, all HTML blocks and inlines must be ignored anyway to get consistent results for all formats. Disabling HTML support completely is also an easy way to avoid a large class of potential security issues.

**Comment by nwellnhof** [comment]
On 04/01/2017 17:29, John MacFarlane wrote:
> The security issue is orthogonal. We have the --safe option
> (and corresponding option for the C API). That will ensure
> that raw HTML doesn't come through and is sufficient for the
> security concern.

Right, but CMARK_OPT_SAFE is a render option that only works if you use one of 
cmark's renderers.

> Having raw HTML parsed as normal characters would be a
> different way to address the security concern, but unlike
> --safe it changes the text -> AST translation in a way that is
> contrary to the spec.

Just like CMARK_OPT_SMART is contrary to the spec.

**Comment by jgm** [comment]
+++ Nick Wellnhofer [Jan 04 17 08:44 ]:

>   Just like CMARK_OPT_SMART is contrary to the spec.

Actually it's not, since it only affects rendering.
The spec is about the text -> AST transformation,
defining the meaning of various elements in the text.

**Comment by jgm** [comment]
+++ Nick Wellnhofer [Jan 04 17 08:44 ]:
>   Right, but CMARK_OPT_SAFE is a render option that only works if you use
>   one of cmark's renderers.

If you use a custom renderer you can always do the sme thing
and ignore raw HTML, right?

**Comment by nwellnhof** [comment]
> If you use a custom renderer you can always do the sme thing and ignore raw HTML, right?

Yes, of course. I'd still find an option to ignore all HTML during parsing convenient for my users. I'm aware that this creates a kind of CommonMark dialect which counters the CommonMark philosophy.

**Comment by foonathan** [comment]
Maybe add a note in the spec that HTML does not need to be supported? Because there clearly are situations were it is not.

**Comment by nwellnhof** [comment]
@foonathan The spec doesn't care whether an output format supports HTML or not. It's only about the text -> AST transformation and ignoring HTML results in a different AST.

**Comment by foonathan** [comment]
I meant HTML *input* like HTML blocks and Raw HTML. They can only conditionally be supported.

**Comment by mity** [comment]
Well, it's maybe more a dispute what Markdown or CommonMark is or what is should be. Historically, it was always more or less tight with web and HTML technologies.

The raw HTML feature more or less anticipates the final consumer of the Markdown is a web browser. Saying otherwise would mean that any fully compliant Markdown renderer must also incorporate full HTML rendering.

On the other side, the practice has already shown Markdown may also be very useful in other applications. Generally speaking, my personal preference would be also "general purpose document format which does not anitcipate much about the renderer side." rather then just "easier way how to write HTML".

In such world, the raw HTML should maybe rather be an extension to Markdown format similar to other extensions comparable to other possible domain specific extensions like e.g. `[[wiki-style link]]`, if you implement a wiki system, or `@username` if you implement a chat application.

(BTW, MD4C does have the option to disable the raw HTML and some other parsers have similar options, likely for exactly the same reason.)

**Comment by jgm** [comment]
+++ Jonathan Müller [Jan 03 17 23:57 ]:
>   Using [foo<T>] is more typing though. And there are other cases where
>   putting a code span would complicate the implementation.

For reasons of consistency and good typography, you should
always use code spans for code, including `foo<T>`.  Then
you don't have to worry about these escaping issues.

>   If you don't want to support it, I have a workaround. But I think there
>   is some use for this option - for example if you intend to output Latex
>   where HTML is just ignored. Then the user can't reasonable input it, so
>   it might jus be treated as text.

I really want to avoid options that affect the creation of
the AST and, in effect, multiply "dialects" of CommonMark.
I don't see a compelling reason to do this, so I'll go ahead
and close this issue.

**Comment by jgm** [comment]
@mity I am sympathetic philosophically.  I don't take an HTML-centric view of CommonMark.
However, CommonMark aims to be a rational reconstruction of Markdown, which has always allowed raw HTML.

**Comment by jgm** [comment]
One more comment on this (and perhaps I'll reopen because maybe this matters).

The default HTML renderer has an option that simply omits raw HTML. But one could alternatively have a renderer option that emits HTML in escaped form.

This would preserve the text -> AST transformation; HTML would still be parsed as raw HTML elements. It would only affect how these elements were rendered. I wouldn't have the same problem with such an option as I would have with an option that affected parsing.

The two options are different.  With a flag affecting parsing, you'd get emphasis in the following example:

    <a href="*emph">*

With a flag affecting rendering, you wouldn't.

**Comment by robinst** [comment]
> But one could alternatively have a renderer option that emits HTML in escaped form.

FWIW, in [commonmark-java](https://github.com/atlassian/commonmark-java/) that's an option for the HTML renderer we provide. So +1 for that.
